examples/seismic/tutorials/05_dask.ipynb

```diff
@@ -0,0 +1,496 @@
+    "    \n",
+    "    # Taking here to avoid a (probably overzealous) destructor deleting\n",
+    "    # the gradient before Dask has had a chance to communicate it.\n",
+    "    g = np.array(grad.data[:])\n",
+    "    \n",
+    "    # return the objective functional and gradient.\n",
+    "    return fg_pair(f, g)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define the global functional-gradient operator. This does the following:\n",
+    "* Creates the Dask cluster\n",
+    "* Maps the worklist (shots) to the cluster so that the invidual contributions to (f, g) are computed.\n",
+    "* Sum individual contributions to (f, g) and returns the result."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from distributed import Client\n",
+    "\n",
+    "# Dumps the model to disk; workers will pick this up when they need it.\n",
+    "def dump_model(param, model):\n",
+    "    np.save(param['model'], model.astype(np.float32))\n",
+    "\n",
+    "def fwi_gradient(model, param):\n",
+    "    # Dump a copy of the current model for the workers\n",
+    "    # to pick up when they are ready.\n",
+    "    param['model'] = \"model_0.npy\"\n",
+    "    dump_model(param, model)\n",
+    "\n",
+    "    # Define work list\n",
+    "    work = [dict(param) for i in range(param['nshots'])]\n",
+    "    for i in  range(param['nshots']):\n",
+    "        work[i]['shot_id'] = i\n",
+    "        \n",
+    "    # Start dask cluster\n",
+    "    client = Client()\n",
+    "    \n",
+    "    # Map worklist to cluster\n",
+    "    fgi = client.map(fwi_gradient_i, work)\n",
+    "    \n",
+    "    # Perform reduction on cluster\n",
+    "    fg = client.submit(sum, fgi).result()\n",
+    "    \n",
+    "    # Shutdown cluster\n",
+    "    client.close()\n",
+    "    \n",
+    "    # L-BFGS in scipy expects a flat array in 64-bit floats.\n",
+    "    return fg.f, fg.g.flatten().astype(np.float64)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## FWI with L-BFGS-B\n",
+    "Equipped with a function to calculate the functional and gradient, we are finally ready to define the optimization function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from scipy import optimize\n",
+    "\n",
+    "# Define bounding box constraints on the solution.\n",
+    "def apply_box_constraint(m):\n",
+    "    # Maximum accepted velocity is 3.5 km/sec (true is 3 km/sec)\n",
+    "    # Minimum accepted velocity is 2 km/sec (true is 2.5 km/sec)\n",
+    "    return np.clip(m, 0.08, 0.25)\n",
+    "\n",
+    "# Many methods in scipy.optimize.minimize allow us to define a callback\n",
```

Comment:
The word method here seems unclear. Does this refer to the programming concept of class methods? Does it refer to an optimization method? 
It seems the sentence above would be clearer if we skip everything before scipy.optimize.minimize above

Reply:
I wouldn't as callback (constraints) are only compatible with part of the optimization methods, not all.

Reply:
Looking at the [documentation] of `scipy.optimize.minimize`, it accepts a callback parameter regardless of which optimization method you choose (which is a different parameter). I don't see any mention that it only works for certain optimization methods. What am I missing?

In any case, since this is not a tutorial of `scipy.optimize.minimize` I think we would be forgiven if we say something that applies to our optimization method of choice but not to every method, for the sake of clarity.

Reply:
sure.

Reply:
changed to:
# Many optimization methods in scipy.optimize.minimize accept a callback
# function that can operate on the solution after every iteration.